Address multiple-parent `:lincomb` in C code generation

mfasi opened the pull request.

jarlebring commented on src/compress_graph.jl:
```diff
@@ -110,24 +111,19 @@ end
 # Check if node is trivial because of an identity on the right (:mult).
 function is_trivial_right(graph, node)
+    return graph.operations[node] == :mult &&
+        graph.parents[node][2] == :I
```
A lincomb can have just one parent. Explicit use of if-statement might be needed.

mfasi replied:
Thank you for spotting this – I've added a new test that should cover this case, but I don't think there is an issue in the case you suggest.

The `&&` operator uses short-circuit evaluation, and if `graph.operations[node]` is not `:mult`, then the second condition is not checked and the operator returns `false` immediately.

jarlebring commented:
Julia code:
```
julia> (g,_)=graph_ps_degopt([2.2,3,4,0,3.3,0.9,0.1]);
julia> gen_code("/tmp/bla.c",g,lang=LangC_MKL());
julia> eval_graph(g,[0 3.0; 0 0])
2×2 Matrix{Float64}:
 2.2  9.0
 0.0  2.2
```
C-code `/tmp/mainfile.c` badly written...  
```
#include<stdio.h>
#include "bla.c"
typedef double blas_type;
int main() {
    size_t i;
    size_t n = 2;
    blas_type *A = malloc(n * n * sizeof(*A));
    for(i = 0; i < n * n; i++){
      A[i] = 0.0;
    }
    A[1]=3.0;
    printf("A\n%lf\n",A[0]);
    printf("%lf\n",A[1]);
    printf("%lf\n",A[2]);
    printf("%lf\n",A[3]);
    blas_type *B = malloc(n * n * sizeof(*A));
    ddummy(A, n, B);
    printf("B\n%lf\n",B[0]);
    printf("%lf\n",B[1]);
    printf("%lf\n",B[2]);
    printf("%lf\n",B[3]);
    return 0;
}
```
bash:
```
jarl@beech:/tmp$ gcc -o main_compiled ./mainfile.c -lmkl_rt 
jarl@beech:/tmp$ ./main_compiled 
A
0.000000
3.000000
0.000000
0.000000
B
2.200000
3.000000
0.000000
2.200000
```


I think the julia and c output are not consistent. Should be 9 instead of 3. Right?

jarlebring commented:
Generated code  

```
   /* Smart lincomb recycle B_1_2. */
    for (size_t i = 0; i < n * n; i += n + 1) {
        *(memslots[2] + i) = coeff_P1_1 * *(memslots[4] + i) +
            coeff_P1_2 * *(memslots[2] + i);
    }
```
Should it really be `i += n + 1` ? We want to do this on all elements, so shouldn't it be `i++` ? For-loops adding multiples of the identity should still have `i += n+1`.

mfasi commented:
Thank you for spotting that bug! Your diagnosis is correct. I introduced that bug when switching to `memset` – I must have copied the wrong line. I've fixed that commit in [`c1e0dc7`](https://github.com/matrixfunctions/GraphMatFun.jl/pull/77/commits/c1e0dc756d262c3776bab24a6da7a9852361eed8) and I've updated the branch.

jarlebring commented:
Shouldn't there be a  `coeff_Bb4_1`? 

```
    double coeff_Bb4_2 = 0.626189194268691;
    double coeff_Bb4_3 = 0.796941757132641;
    double coeff_Bb4_4 = 0.21311590849480466;
    for (size_t i = 0; i < n * n; i++) {
        *(memslots[3] + i) = coeff_Bb4_2 * *(memslots[0] + i) + 
            coeff_Bb4_3 * *(memslots[1] + i) + 
            coeff_Bb4_4 * *(memslots[2] + i);
    }
    double coeff_Bb4_0 = 0.5728465547708623;
    for (size_t i = 0; i < n * n; i += n + 1) {
        *(memslots[3] + i) += coeff_Bb4_0;
    }
```

jarlebring commented:
For me, the generated main for complex numbers does not work.

```
julia> (g,_)=graph_ps_degopt(rand(10)+1im*rand(10));
julia> gen_code("/tmp/bla.c",g,lang=LangC_MKL(true));
```

```
 jarl@beech:/tmp$ gcc -o main_compiled /tmp/bla.c -lmkl_rt
/tmp/bla.c: In function ‘main’:
/tmp/bla.c:208:16: error: incompatible types when assigning to type ‘blas_type’ {aka ‘struct _MKL_Complex16’} from type ‘double’
  208 |         A[i] = rand() / (1.0 * RAND_MAX);
```

mfasi commented:
> Shouldn't there be a `coeff_Bb4_1`?
> 
> ```
>     double coeff_Bb4_2 = 0.626189194268691;
>     double coeff_Bb4_3 = 0.796941757132641;
>     double coeff_Bb4_4 = 0.21311590849480466;
>     for (size_t i = 0; i < n * n; i++) {
>         *(memslots[3] + i) = coeff_Bb4_2 * *(memslots[0] + i) + 
>             coeff_Bb4_3 * *(memslots[1] + i) + 
>             coeff_Bb4_4 * *(memslots[2] + i);
>     }
>     double coeff_Bb4_0 = 0.5728465547708623;
>     for (size_t i = 0; i < n * n; i += n + 1) {
>         *(memslots[3] + i) += coeff_Bb4_0;
>     }
> ```

Currently, I'm using as index the position of the matrix as parent in the `lincomb` node, and using the special index `0` to accumulate the coefficients of all the identities. I can change this to use a counter that is incremented every time a new coefficient is found – will do.

> For me, the generated main for complex numbers does not work.
> 
> ```
> julia> (g,_)=graph_ps_degopt(rand(10)+1im*rand(10));
> julia> gen_code("/tmp/bla.c",g,lang=LangC_MKL(true));
> ```
> 
> ```
>  jarl@beech:/tmp$ gcc -o main_compiled /tmp/bla.c -lmkl_rt
> /tmp/bla.c: In function ‘main’:
> /tmp/bla.c:208:16: error: incompatible types when assigning to type ‘blas_type’ {aka ‘struct _MKL_Complex16’} from type ‘double’
>   208 |         A[i] = rand() / (1.0 * RAND_MAX);
> ```

I don't think this has ever worked for complex types with `LangC_MKL()`, because assigning a single floating-point number is not valid with MKL's complex types. This PR is probably a good chance to fix this, however.

jarlebring commented on src/code_gen/gen_c_code.jl:
```diff
@@ -107,10 +94,97 @@ function get_blas_type(::LangC_OpenBLAS, T::Type{Complex{Float64}})
+    acc->imag += *a * b->imag;
+}
+
+void prod_MKL_Complex8_float(MKL_Complex8 *acc,
```
Maybe include comments what `prod_`,  `acc_` etc mean?

mfasi commented:
It looks like a bug – I'll try to debug it more carefully.

One unrelated comment is that the matrices should be contiguous by columns rather than by row (the BLAS and LAPACK routines use `CblasColMajor` and `LAPACK_COL_MAJOR`, respectively), so `A[1]` and `A[2]` should be swapped.

mfasi commented:
Found the bug. The two functions:

```
// Compute c <- c + a * b.
void fma_MKL_Complex16(MKL_Complex16 *c,
                    const MKL_Complex16 *a,
                    const MKL_Complex16 *b) {
    c->real += a->real * b->real - a->imag * b->imag;
    c->imag += a->real * b->imag + a->imag * b->real;
}

// Compute c <- a * b.
void prod_MKL_Complex16(MKL_Complex16 *c,
                    const MKL_Complex16 *a,
                    const MKL_Complex16 *b) {
    c->real = a->real * b->real - a->imag * b->imag;
    c->imag = a->real * b->imag + a->imag * b->real;
}
```
should really be
```
// Compute c <- c + a * b.
void fma_MKL_Complex16(MKL_Complex16 *c,
                    const MKL_Complex16 *a,
                    const MKL_Complex16 *b) {
    double real_part = a->real * b->real - a->imag * b->imag;
    double imag_part = a->real * b->imag + a->imag * b->real;
    c->real += real_part;
    c->imag += imag_part;
}

// Compute c <- a * b.
void prod_MKL_Complex16(MKL_Complex16 *c,
                    const MKL_Complex16 *a,
                    const MKL_Complex16 *b) {
    double real_part = a->real * b->real - a->imag * b->imag;
    double imag_part = a->real * b->imag + a->imag * b->real;
    c->real = real_part;
    c->imag = imag_part;
}
```
because `c` can be the same as `a` or `b` if we are reusing memory.

Fixed in my latest commit.

jarlebring commented on src/code_gen/gen_c_code.jl:
```diff
@@ -589,22 +626,22 @@ function gen_main(lang::LangC, T, fname, funname; A = 10::Union{Integer,Matrix})
-            push_code!(code, "A[i] = rand() / (1.0*RAND_MAX);", ind_lvl = 2)
+            push_code!(code, "blas_type *A = malloc(n * n * sizeof(*A));")
+            push_code!(code, "for(i = 0; i < n * n; i++){")
+            push_code!(code, "A[i] = rand() / (1.0 * RAND_MAX);", ind_lvl = 2)
```
separate between complex and real cases? 
```
if eltype(g) <: Complex 
            push_code!(code, "A[i].real = rand() / (1.0 * RAND_MAX);", ind_lvl = 2)
            push_code!(code, "A[i].imag = rand() / (1.0 * RAND_MAX);", ind_lvl = 2)
else
            push_code!(code, "A[i] = rand() / (1.0 * RAND_MAX);", ind_lvl = 2)
end
```

mfasi replied:
> separate between complex and real cases?
> 
> ```
> if eltype(g) <: Complex 
>             push_code!(code, "A[i].real = rand() / (1.0 * RAND_MAX);", ind_lvl = 2)
>             push_code!(code, "A[i].imag = rand() / (1.0 * RAND_MAX);", ind_lvl = 2)
> else
>             push_code!(code, "A[i] = rand() / (1.0 * RAND_MAX);", ind_lvl = 2)
> end
> ```

Had to do this in a slightly roundabout way because complex types are not assigned in the same way in MKL and OpenBLAS, but this is the idea I implemented.

mfasi commented:
> lapack_complex_double

It works with both `openblas_complex_double` and `lapack_complex_double`, so I suggest sticking to the latter if that means it will work for you as well. I've made the change.

mfasi commented:
> https://github.com/matrixfunctions/GraphMatFun.jl/blob/fac0e489812b61d02b3caf04087040ceab554c53/src/code_gen/gen_c_code.jl#L623-L626
> 
> As far as I understand, there is no way for a user to influence kwarg A. An end-user anyway shouldn't directly call `gen_main`. Maybe we can just remove support for what is not default. That way we can also remove `print_indented_matrix`.
> 
> The alternative is to expose it to the user through a parameter in the `LangC`-type, but it seems a bit much for. "The best code is no code"
> 
> This is actually already present before this PR, but when we are anyway messing with this code...

Indeed I was a bit puzzled by the kwarg that cannot be used – there must have been a way at some point in the past, but I don't remember how we did it. I've removed `A` and all related functionalities (which were not – and I believe could not – be covered by our tests).

jarlebring commented:
Great! LGTM 👍

mfasi commented:
One other thing I noticed by looking at the generated code is that we now have a number of new trivial `lincomb` nodes, such as
```
    X ← + α × A + 0 × B
```
which should really be
```
    X ← + α × A             (a)
```
or
```
X ← + 1 × A
```
which is just a copy and should be removed. In fact, nodes of the form `(a)` could also be removed by pushing `α` to the corresponding coefficient(s) of sufficient `lincomb`s, although this seems a bit messier to do.

mfasi commented:
> Great! LGTM 👍

Great! I'll create an issue from my last comment and let's discuss it there.

jarlebring commented:
> 
> ```
>     X ← + α × A + 0 × B
> ```
> 
> which should really be
> 
> ```
>     X ← + α × A             (a)
> ```
> 
> or
> 

I actually think we this type of code optimization should not be done in the code generation. A user can always run `compress_trivial!` before calling `gen_code` and those cases will be removed. It can be handled at graph-level. I think we get cleaner code-generation code if we see the code generator as being a bit stupid. Code generation philosophy: "Generate code for exactly this graph".
